app/src/testDebug/java/com/futurice/freesound/feature/common/ui/adapter/RecyclerViewAdapterTest.java

```diff
@@ -0,0 +1,212 @@
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+
+        SchedulerProvider schedulerProvider = new TrampolineSchedulerProvider();
+        adapter = new RecyclerViewAdapter(comparator, factoryMap(), binderMap(), schedulerProvider);
+    }
+
+    @Test
+    public void expectedNPEWhenViewGroupIsNullInOnCreateViewHolder() {
+        thrown.expect(NullPointerException.class);
+
+        adapter.onCreateViewHolder(null, 1);
+    }
+
+    @Test
+    public void expectedNPEWhenTypeDoesNotExistInOnCreateViewHolder() {
+        thrown.expect(NullPointerException.class);
+
+        adapter.onCreateViewHolder(null, 5);
+    }
+
+    @Test
+    public void factoryForPassedTypeCreatesViewHolder() {
+        ViewGroup viewGroup = mock(ViewGroup.class);
+
+        adapter.onCreateViewHolder(viewGroup, 1);
+
+        Mockito.verify(factory1).createViewHolder(viewGroup);
+    }
+
+    @Test
+    public void expectedNPEWhenNullIsPassedToUpdate() {
+        thrown.expect(NullPointerException.class);
+
+        adapter.update(null);
+    }
+
+    @Test
+    public void diffUtilsIsNotUsedWhenUpdatingAdapterForTheFirstTime() {
+        List<DisplayableItem> itemList = itemList();
+
+        // FIXME: AndroidPreconditions.assertUiThread(), how to mock this ?
+        adapter.update(itemList);
+
+        verify(comparator, never()).areItemsTheSame(any(), any());
+    }
+
+    @Test
+    public void diffUtilsIsUsedWhenUpdatingAdapterForTheFirstTime() {
+        List<DisplayableItem> itemList = itemList();
+
+        adapter.update(itemList);
+        adapter.update(itemList);
+
+        verify(comparator, atLeast(1)).areItemsTheSame(any(), any());
+    }
+
+    @Test
+    public void expectedNPEWhenViewHolderIsNullInOnBindViewHolder() {
+        thrown.expect(NullPointerException.class);
+        List<DisplayableItem> itemList = itemList();
+
+        adapter.update(itemList);
+        adapter.onBindViewHolder(null, 1);
+    }
+
+    @Test
+    public void expectedIndexOutOfBoundsExceptionWhenPositionIsOutOfBoundsInOnBindViewHolder() {
+        thrown.expect(IndexOutOfBoundsException.class);
+        List<DisplayableItem> itemList = itemList();
+        RecyclerView.ViewHolder viewHolder = mock(RecyclerView.ViewHolder.class);
+
+        adapter.update(itemList);
+        adapter.onBindViewHolder(viewHolder, 5);
+    }
+
+    @Test
+    public void binderForItemTypeBindsItemToViewHolder() {
+        List<DisplayableItem> itemList = itemList();
+        RecyclerView.ViewHolder viewHolder = mock(RecyclerView.ViewHolder.class);
+
+        adapter.update(itemList);
+        adapter.onBindViewHolder(viewHolder, 1);
+
+        verify(binder2).bind(viewHolder, itemList.get(1));
+    }
+
+    @Test
+    public void unbindIsCalledForBindingViewHolderInOnViewRecycled() {
+        BaseBindingViewHolder viewHolder = mock(BaseBindingViewHolder.class);
+
+        //FIXME: This throws NPE
+        adapter.onViewRecycled(viewHolder);
```

Comment:
The unbind() method in the BindingViewHolder throws a NPE because it is a mock. I need it to be a mock to be able to verify that the unbind() method is called. Any ideas how to test this?